### PR TITLE
fix: update settings observables to detect changes

### PR DESCRIPTION
**Pull request by hkang1**
## Description

Fix issue of log viewer filters not working.
Fix issue of log viewer filter reset not working.
Fix issue of log viewer reset button not showing when search text exists.

## Test Plan

- [ ] navigate to trial or task logs and verify that changing the log filters update the logs.
- [ ] on the same trial/task logs page, apply some filters, verify the reset button appears and clicking it resets the filters.
- [ ] verify when entering log search, reset button appears.
- [ ] There may be other areas that was previously impacted this `useSettings` issue. Other drop downs such as list filters and might be worth checking out as well.

## Commentary (optional)

Looks like the culprit was a `useState` wrapping around an observable. Changes were referencing the state object, which doesn't change when the contents of the observable changes. The `memoized` state was updated to remain an observable and using `useObservable` directly (without `useState`) to point to the state directly so changes are detected properly.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
WEB-1129
WEB-1134

### CODE REVIEW

**Review by keita-determined — COMMENTED** [review]
let me clarify here
Does this fix include `reset` button and `text search` bugs or just dropdown bug?
if its just dropdown, it looks working, but other bugs are not fixed. 
tested on private window (chrome)

https://user-images.githubusercontent.com/109113616/230479220-5fc95d8e-4e47-4bee-bfae-52100fa80fc7.mov

**Comment by hkang1** [comment]
> let me clarify here Does this fix include `reset` button and `text search` bugs or just dropdown bug? if its just dropdown, it looks working, but other bugs are not fixed. tested on private window (chrome)

This just addressed the dropdown but took a look at the reset button issue and the typing text too fast issue. I was able to resolve the reset issue with another commit, but the typing too fast will be better left as a separate PR.

File: webui/react/src/hooks/useSettings.ts [hunk 1]
```diff
@@ -258,27 +252,32 @@ const useSettings = <T>(config: SettingsConfig<T>): UseSettingsReturn<T> => {
...
 
-      stateOb.update((s) => {
-        const news = s.get(config.storagePath);
-        array.forEach((setting) => {
-          let defaultSetting: SettingsConfigProp<T[Extract<keyof T, string>]> | undefined =
-            undefined;
+      stateOb.update((prev) =>
+        prev.withMutations((map) => {
+          const oldSettings = map.get(config.storagePath) ?? {};
+          const newSettings = { ...oldSettings };
```
Comment (keita-determined) [inline webui/react/src/hooks/useSettings.ts:258]: nit: looks like `oldSettings` is not used other than assigning
 
```suggestion
          const newSettings = map.get(config.storagePath) ?? {};
```

**Review by keita-determined — APPROVED** [review]
LGTM

File: webui/react/src/hooks/useSettings.ts (hunk 1 above)
Comment (hkang1) [reply webui/react/src/hooks/useSettings.ts]: Ah yeah, I can probably get rid of `oldSettings` but would still need to make sure that it's a new object to ensure the change is detected.

File: webui/react/src/hooks/useSettings.ts [hunk 2]
```diff
@@ -142,17 +142,19 @@ const useSettings = <T>(config: SettingsConfig<T>): UseSettingsReturn<T> => {
     clearQuerySettings,
   } = useContext(UserSettings);
   const initialLoading = useObservable(isLoadingOb);
-  const [derivedOb] = useState(stateOb.select((s) => s.get(config.storagePath)));
-  const state = useObservable(derivedOb);
+  const settingsByKey = stateOb.select((s) => s.get(config.storagePath));
```
Comment (ClaireNeveu) [inline webui/react/src/hooks/useSettings.ts:145]: This creates a new observable every time the render function is called which orphans any existing subscriptions.

File: webui/react/src/hooks/useSettings.ts [hunk 3]
```diff
@@ -319,7 +317,7 @@ const useSettings = <T>(config: SettingsConfig<T>): UseSettingsReturn<T> => {
       const url = mappedSettings ? `?${mappedSettings}` : '';
       navigate(url, { replace: true });
     });
-  }, [derivedOb, navigate, config, updateDB, initialLoading, loadableUser]);
+  }, [navigate, config, updateDB, initialLoading, loadableUser, settingsByKey]);
```
Comment (ClaireNeveu) [inline webui/react/src/hooks/useSettings.ts:320]: You're creating a new `settingsByKey` every render so this just creates endless new subscriptions meaning an ever increasing number of API calls.

**Pull request merged**